crypto_kem/hqc-128-1-cca2/avx2_leaktime/fft.c

```diff
@@ -0,0 +1,631 @@
+ * https://binary.cr.yp.to/mcbits-20130616.pdf
+ *
+ * @param[out] f Array of size a power of 2
+ * @param[in] f0 Array half the size of f
+ * @param[in] f1 Array half the size of f
+ * @param[in] m_f 2^{m_f} is the smallest power of 2 greater or equal to the number of coefficients of f
+ */
+static void radix_t(uint16_t *f, const uint16_t *f0, const uint16_t *f1, uint32_t m_f) {
+    switch (m_f) {
+    case 4:
+        f[0] = f0[0];
+        f[1] = f1[0];
+        f[2] = f0[1] ^ f1[0];
+        f[3] = f[2] ^ f1[1];
+        f[4] = f[2] ^ f0[2];
+        f[5] = f[3] ^ f1[2];
+        f[6] = f[4] ^ f0[3] ^ f1[2];
+        f[7] = f[3] ^ f0[3] ^ f1[3];
+        f[8] = f[4] ^ f0[4];
+        f[9] = f[5] ^ f1[4];
+        f[10] = f[6] ^ f0[5] ^ f1[4];
+        f[11] = f[7] ^ f0[5] ^ f1[4] ^ f1[5];
+        f[12] = f[8] ^ f0[5] ^ f0[6] ^ f1[4];
+        f[13] = f[7] ^ f[9] ^ f[11] ^ f1[6];
+        f[14] = f[6] ^ f0[6] ^ f0[7] ^ f1[6];
+        f[15] = f[7] ^ f0[7] ^ f1[7];
+        return;
+
+    case 3:
+        f[0] = f0[0];
+        f[1] = f1[0];
+        f[2] = f0[1] ^ f1[0];
+        f[3] = f[2] ^ f1[1];
+        f[4] = f[2] ^ f0[2];
+        f[5] = f[3] ^ f1[2];
+        f[6] = f[4] ^ f0[3] ^ f1[2];
+        f[7] = f[3] ^ f0[3] ^ f1[3];
+        return;
+
+    case 2:
+        f[0] = f0[0];
+        f[1] = f1[0];
+        f[2] = f0[1] ^ f1[0];
+        f[3] = f[2] ^ f1[1];
+        return;
+
+    case 1:
+        f[0] = f0[0];
+        f[1] = f1[0];
+        return;
+
+    default:
+        ;
+
+        size_t n = 1 << (m_f - 2);
+
+        uint16_t Q0[1 << (PARAM_FFT_T - 2)];
+        uint16_t Q1[1 << (PARAM_FFT_T - 2)];
+        uint16_t R0[1 << (PARAM_FFT_T - 2)];
+        uint16_t R1[1 << (PARAM_FFT_T - 2)];
+
+        uint16_t Q[1 << 2 * (PARAM_FFT_T - 2)];
+        uint16_t R[1 << 2 * (PARAM_FFT_T - 2)];
+
+        memcpy(Q0, f0 + n, 2 * n);
+        memcpy(Q1, f1 + n, 2 * n);
+        memcpy(R0, f0, 2 * n);
+        memcpy(R1, f1, 2 * n);
+
+        radix_t (Q, Q0, Q1, m_f - 1);
+        radix_t (R, R0, R1, m_f - 1);
+
+        memcpy(f, R, 4 * n);
+        memcpy(f + 2 * n, R + n, 2 * n);
+        memcpy(f + 3 * n, Q + n, 2 * n);
+
+        for (size_t i = 0 ; i < n ; ++i) {
+            f[2 * n + i] ^= Q[i];
+            f[3 * n + i] ^= f[2 * n + i];
+        }
+    }
+}
+
+
+
+/**
+ * @brief Recursively computes syndromes of family w
+ *
+ * This function is a subroutine of the function fft_t
+ *
+ * @param[out] f Array receiving the syndromes
+ * @param[in] w Array storing the family
+ * @param[in] f_coeffs Length of syndromes vector
+ * @param[in] m 2^m is the smallest power of 2 greater or equal to the length of family w
+ * @param[in] m_f 2^{m_f} is the smallest power of 2 greater or equal to the length of f
+ * @param[in] betas FFT constants
+ */
+static void fft_t_rec(uint16_t *f, const uint16_t *w, size_t f_coeffs,
+                      uint32_t m, uint32_t m_f, const uint16_t *betas) {
+    size_t k = 1 << (m - 1);
```

Comment:
There is a `clang-tidy` warning on this line.

Reply:
I removed the avx2 implementation.